packages/core/crates/core-crypto/src/routing.rs

```diff
@@ -0,0 +1,381 @@
+        let last_hop_len = additional_data_last_hop.map(|d| d.len()).unwrap_or(0) + 1; // end prefix length
+
+        let header_len = last_hop_len + (max_hops - 1) * routing_info_len;
+        let extended_header_len = last_hop_len + max_hops * routing_info_len;
+
+        let mut extended_header = vec![0u8; extended_header_len];
+        let mut ret = RoutingInfo::default();
+
+        for idx in 0..secrets.len() {
+            let inverted_idx = secrets.len() - idx - 1;
+            let secret = secrets[inverted_idx];
+            let prg = PRG::from_parameters(PRGParameters::new(secret));
+
+            if idx == 0 {
+                extended_header[0] = RELAYER_END_PREFIX;
+
+                if let Some(data) = additional_data_last_hop {
+                    extended_header[1..data.len()].copy_from_slice(data);
+                }
+
+                let padding_len = (max_hops - secrets.len()) * routing_info_len;
+                if padding_len > 0 {
+                    random_fill(&mut extended_header[last_hop_len..padding_len]);
+                }
+
+                let key_stream = prg.digest(0, last_hop_len + padding_len);
+                xor_inplace(&mut extended_header[0..last_hop_len + padding_len], &key_stream);
+
+                if secrets.len() > 1 {
+                    let filler = generate_filler(max_hops, routing_info_len, last_hop_len, secrets);
+                    extended_header[last_hop_len + padding_len..last_hop_len + padding_len + filler.len()]
+                        .copy_from_slice(&filler);
+                }
+            } else {
+                extended_header.copy_within(0..header_len, routing_info_len);
+                extended_header[0..PublicKey::SIZE_COMPRESSED].copy_from_slice(&path[inverted_idx + 1].to_bytes(true));
+                extended_header[PublicKey::SIZE_COMPRESSED..PublicKey::SIZE_COMPRESSED + SimpleMac::SIZE]
+                    .copy_from_slice(&ret.mac);
+
+                extended_header[PublicKey::SIZE_COMPRESSED + SimpleMac::SIZE
+                    ..PublicKey::SIZE_COMPRESSED + SimpleMac::SIZE + additional_data_relayer[inverted_idx].len()]
+                    .copy_from_slice(additional_data_relayer[inverted_idx]);
+
+                let key_stream = prg.digest(0, header_len);
+                xor_inplace(&mut extended_header, &key_stream);
+            }
+
+            let mut m = derive_mac_key(secret).and_then(|k| SimpleMac::new(&k)).unwrap();
+            m.update(&extended_header[0..header_len]);
+            m.finalize_into(&mut ret.mac);
+        }
+
+        ret.routing_information = (&extended_header[0..header_len]).into();
+        ret
+    }
+}
+
+/// Enum carry information about the packet based on whether it is destined for the current node (`FinalNode`)
+/// or if the packet is supposed to be only relayed (`RelayNode`).
+pub enum ForwardedHeader {
+    /// Packet is supposed to be relayed
+    RelayNode {
+        /// Transformed header
+        header: Box<[u8]>,
```

Comment:
The problem with having non-native objects is that each Box is an allocation operation. For a large enough effort, allocations might become as expensive as crypto operations for the packet construction. Given a fixed size of the packet, it should not be necessary to perform any allocations what so ever.

Reply:
I've experimented with that first, but resulted in object that would be really tough to make work with `wasm_bindgen`.

Reply:
But does the packet still have to be moved across the WASM boundary?